Sources/NIOOpenSSL/SSLConnection.swift

```diff
@@ -81,6 +68,16 @@ internal final class SSLConnection {
         SSL_set_connect_state(ssl)
     }
 
+    func setAllocator(_ allocator: ByteBufferAllocator) {
+        self.bio = ByteBufferBIO(allocator: allocator)
+
+        // This weird dance where we pass the *exact same* pointer in to both objects is because, weirdly,
+        // the OpenSSL docs claim that only one reference count will be consumed here. We therefore need to
+        // avoid calling BIO_up_ref too many times.
+        let bioPtr = self.bio!.bio
+        SSL_set_bio(self.ssl, bioPtr, bioPtr)
```

Comment:
and the previous `bioPtr` will be `down_ref`'d automatically? what happens if old `bioPtr` and new `bioPtr` are equal, does openssl handle that too?

Reply:
This function is some bullshit. Even the manual acknowledges it. From the man page:

> `SSL_set_bio()` does a similar job as `SSL_set0_rbio()` and `SSL_set0_wbio()` except that it connects both the `rbio` and the `wbio` at the same time. This function transfers the ownership of `rbio` and `wbio` to `ssl` except that the rules for this are much more complex. For this reason this function is considered a legacy function and `SSL_set0_rbio()` and `SSL_set0_wbio()` should be used in preference. The ownership rules are as follows:
>
> - If neither the `rbio` or `wbio` have changed from their previous values then nothing is done.
> - If the `rbio` and `wbio` parameters are different and both are different to their previously set values then one reference is consumed for the `rbio` and one reference is consumed for the `wbio`.
> - If the `rbio` and `wbio` parameters are the same and the `rbio` is not the same as the previously set value then one reference is consumed.
> - If the `rbio` and `wbio` parameters are the same and the `rbio` is the same as the previously set value, then no additional references are consumed.
> - If the `rbio` and `wbio` parameters are different and the `rbio` is the same as the previously set value then one reference is consumed for the `wbio` and no references are consumed for the `rbio`.
> - If the `rbio` and `wbio` parameters are different and the `wbio` is the same as the previously set value and the old `rbio` and `wbio` values were the same as each other then one reference is consumed for the `rbio` and no references are consumed for the `wbio`.
> - If the `rbio` and `wbio` parameters are different and the `wbio` is the same as the previously set value and the old `rbio` and `wbio` values were different to each other then one reference is consumed for the `rbio` and one reference is consumed for the `wbio`.

*Unfortunately*, the replacement functions are OpenSSL 1.1 only, and it is deeply non-trivial to try to use `SSL_set_bio` to re-implement the two replacement functions. In this case, we just rely on the documented OpenSSL behaviour and the fact that all of this code is internal.

Reply:
ok cool, then we need an `assert(bioPtr != getOldBIOPtr())`